packages/whenable/src/watch.js

```diff
@@ -0,0 +1,239 @@
+// @ts-check
+import { M } from '@endo/patterns';
+
+import { getWhenablePayload, unwrapPromise, basicE } from './whenable-utils.js';
+
+const { Fail } = assert;
+
+const { apply } = Reflect;
+
+export const PromiseWatcherI = M.interface('PromiseWatcher', {
+  onFulfilled: M.call(M.any()).rest(M.any()).returns(),
+  onRejected: M.call(M.any()).rest(M.any()).returns(),
+});
+
+/**
+ * @typedef {object} PromiseWatcher
+ * @property {(...args: unknown[]) => void} [onFulfilled]
+ * @property {(...args: unknown[]) => void} [onRejected]
+ */
+
+/**
+ * Adapt a promise watcher method to E.when.
+ * @param {Record<PropertyKey, (...args: unknown[]) => unknown>} that
+ * @param {PropertyKey} prop
+ * @param {unknown[]} postArgs
+ */
+const callMeMaybe = (that, prop, postArgs) => {
+  const fn = that[prop];
+  if (!fn) {
+    return undefined;
+  }
+  assert.typeof(fn, 'function');
+  /**
+   * @param {unknown} arg value or reason
+   */
+  const wrapped = arg => {
+    // Don't return a value, to prevent E.when from subscribing to a resulting
+    // promise.
+    apply(fn, that, [arg, ...postArgs]);
+  };
+  return wrapped;
+};
+
+/**
+ * Shim the promise watcher behaviour when VatData.watchPromise is not available.
+ *
+ * @param {Promise<any>} p
+ * @param {PromiseWatcher} watcher
+ * @param {...unknown[]} watcherArgs
+ * @returns {void}
+ */
+const watchPromiseShim = (p, watcher, ...watcherArgs) => {
+  Promise.resolve(p) === p || Fail`watchPromise only watches promises`;
```

Comment:
I don't understand the connection between the condition being asserted and the error message. No change suggested. I'm just asking for my own clarity.

Reply:
Nevermind. Resolving and then deleting. Sorry for the noise.